Homework #8. Tic Tac Toe. Волынец Арсений

vol0n opened the pull request.

vol0n committed:
First working version

vol0n committed:
Fix detekt

vol0n committed:
Try to fix build on windows

SpirinEgor commented on src/main/kotlin/hw8/controllers/GameController.kt:
```diff
@@ -0,0 +1,32 @@
+
+    init {
+        GameModel.onEndGame = { player, _ ->
+            val resMes = when (player) {
+                null -> "It is draw: nobody won!"
+                else -> "${player.name} won!"
+            }
```
Еще можно `val resultMessage = "${player?.name} won!" ?: "It is a draw: nobody won!"`
Ну и да `resMes` не лучшее название

vol0n replied:
вроде так нельзя: Elvis operator (?:) always returns the left operand of non-nullable type String

SpirinEgor commented on src/main/kotlin/hw8/model/GameModel.kt:
```diff
@@ -0,0 +1,159 @@
+    const val EMPTY = ""
+    const val boardSize = 3
+
+    private val board: List<MutableList<SimpleStringProperty>> = List(this.boardSize) { rowIndex ->
```
Как так вышло, что модель, которая отвечает за логику, знает где и как это поле будет отображаться?

vol0n replied:
не очень понимаю, что модель знает про то где и как поле будет отображаться, надо было сделать String вместо SimpleStringProperty или вообще Int? Модель же в любом случае должна предоставлять какое-то api, что у нее поменялась такая-то клетка на поле, я подумал что удобно использовать Property

SpirinEgor commented on src/main/kotlin/hw8/model/GameModel.kt:
```diff
@@ -0,0 +1,159 @@
+    private var gameStarted = false
+
+    fun addPlayer(player: Player) {
+        if (!gameStarted) {
+            players.add(player)
+        }
```
Так до бесконечности игроков можно добавлять?) Вроде как в крестиках-ноликах есть инвариант в 2 игрока

SpirinEgor commented on src/main/kotlin/hw8/model/GameModel.kt:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    // for complete restart
+    fun clearPlayersAndRestart() {
+        restart()
+        players.clear()
+    }
```
Переочищать не выглядит хорошим решением. Если будут появляться новые сущности надо будет не забывать переинициализировать их тут. Гораздо лучше избавиться от object и сделать нормальный класс. Модель будет создаваться одна на игру со всеми нужными полями.

SpirinEgor commented on src/main/kotlin/hw8/model/Player.kt:
```diff
@@ -0,0 +1,59 @@
+    override val isLongResponse = false
+    override fun retrieveTurn(): Turn {
+        val candidates = mutableListOf<Turn>()
+        GameModel.gameBoard.mapIndexed { i, row ->
```
Ну это совсем жесть. Лезть в какие-то глобальные состояния, чтобы узнать положение. Я уже вижу этот CheaterPlayer, который меняет это глобальное состояние, чтобы побеждать)

vol0n replied:
Ну там же GameModel.gameBoard возвращает ReadOnlyStringProperty, а их нельзя менять, поэтому читер не пройдет вроде)

SpirinEgor commented on src/main/kotlin/hw8/views/MenuScreen.kt:
```diff
@@ -0,0 +1,102 @@
+    fun userMenuMessage(): String
+}
+
+object RandomMode : GameMode {
```
Что же ты так эти object любишь. Делай нормальные классы)

vol0n replied:
Хорошо)

SpirinEgor commented on src/main/kotlin/hw8/views/MenuScreen.kt:
```diff
@@ -0,0 +1,102 @@
+import tornadofx.vbox
+
+interface GameMode {
+    fun createPlayer(): Player
```
И тут тоже совсем нелогично. У тебя вдруг представления начали сами создавать игроков.

vol0n replied:
тут я тоже не очень понял, надо переместить это в другой файл? Или нелогично, что представление создает игроков (в методе onUndock()), а этим лучше должен заниматься контроллер? Или это все сразу?)

SpirinEgor reviewed (changes requested):
Игра работает, выглядит хорошо. Единственное, когда выбираю игру за нолики, то все равно хожу первым...

По реализации -- супер запутанного кода нет, но текущая реализация далека от MVC, кажется это будет сложно масштабировать на новые режимы/возможности. Ну и идея, что поле это какой-то глобальный одиночный режим просто ужас)

vol0n committed:
Use classes not objects, use enums not Properties

- GameModel is now a class not an object, the same goes for Player
  concrete classes.
- The board property of GameModel is now using enum class Cell
- Some initialization logic was moved from views to controller.
- Restart methods were removed, now for new game there will be created
  new GameModel instance.
- Now the first move is done by player with X label, and the second
move is done by a player with O label.

vol0n committed:
Move creating players logic from controller to model

SpirinEgor commented on src/main/kotlin/hw8/model/GameModel.kt:
```diff
@@ -0,0 +1,155 @@
+        val playersLabels = listOf(Cell.X, Cell.O)
+    }
+
+    protected abstract val players: MutableList<Player>
```
А зачем мутабельным делать?

vol0n replied:
чтобы можно порядок игроков менять, мне почему-то не хотелось порядок через конструктор передавать, наверное так все же лучше, а список сделать неизменяемым?

SpirinEgor replied:
Ну у тебя на каждую желательно отдельный объект заводить, а не переиспользовать имеющийся. А так вообще заводить список под 2 игроков, мб проще поддерживать 2 переменные и совать их по необходимости

SpirinEgor commented on src/main/kotlin/hw8/model/Player.kt:
```diff
@@ -0,0 +1,58 @@
+    val name: String
+}
+
+class ComputerRandom(private val game: GameModel) : Player {
+    companion object {
+        fun chooseRandomTurn(game: GameModel): Turn {
```
А в чем прикол 2 раза модель игры передавать? Да и вообще кажется, что должно быть наоборот -- игра знает про игроков. Ну то есть реален читер, который может обращаться напрямую к игре и подменять поле.

SpirinEgor reviewed (changes requested):
Ага, теперь выглядит гораздо лучше, хоть и есть несколько архитектурных замечаний. Теперь это даже можно адекватно расширить на сетевой режим -- реализовать Player, который ходит на сервер за ходом противника. Не самое лучшее решение, но рабочее и в домашке сойдет.

vol0n committed:
First working version online mode

vol0n committed:
Fix detekt issues

vol0n committed:
Add notification when opponent left the game

- Server keeps track of active games.
- Now when game is live and opponent has disconnected,
WebGame will send to server the Disconnected message.
Server will find the active game and send message to
another player.

vol0n committed:
Add back to mode menu button

vol0n committed:
Add config for client

vol0n committed:
Try to fix build

vol0n committed:
Fix conflicts

vol0n committed:
Update library versions, fix test in 3 hw

vol0n committed:
Fix build: use replaceFistChar in hw3 test

SpirinEgor reviewed (approved):
Очень классно вышло, адекватная таблица игроков и возможность посылать приглашения. Можешь потом в отдельный репо вынести и хвастаться :)

nbirillo reviewed (approved):
Да, мне в целом тоже все нравится
